Python/ast_opt.c

```diff
@@ -125,6 +125,132 @@ fold_unaryop(expr_ty node, PyArena *arena)
     return make_const(node, newval, arena);
 }
 
+/* Check whether a collection doesn't containing too much items (including
+   subcollections).  This protects from creating a constant that needs
+   too much time for calculating a hash.
+   "limit" is the maximal number of items.
+   Returns the negative number if the total number of items exceeds the
+   limit.  Otherwise returns the limit minus the total number of items.
+*/
+
+static Py_ssize_t
+check_complexity(PyObject *obj, Py_ssize_t limit)
+{
+    if (PyTuple_Check(obj)) {
+        Py_ssize_t i;
+        limit -= PyTuple_GET_SIZE(obj);
+        for (i = 0; limit >= 0 && i < PyTuple_GET_SIZE(obj); i++) {
+            limit = check_complexity(PyTuple_GET_ITEM(obj, i), limit);
+        }
+        return limit;
+    }
+    else if (PyFrozenSet_Check(obj)) {
+        Py_ssize_t i = 0;
+        PyObject *item;
+        Py_hash_t hash;
+        limit -= PySet_GET_SIZE(obj);
+        while (limit >= 0 && _PySet_NextEntry(obj, &i, &item, &hash)) {
+            limit = check_complexity(item, limit);
+        }
+    }
+    return limit;
+}
+
+#define MAX_INT_SIZE 128
+#define MAX_COLLECTION_SIZE 256
+#define MAX_STR_SIZE 4096
+#define MAX_TOTAL_ITEMS 1000000
```

Comment:
Wow, that's a lot. What's the rationale for so big objects? The old peephole optimizer limited to tuples of 20 items or something like that, no?

Reply:
The old peephole optimizer doesn't have such limit (see bpo-30416). This is a limit for the total number of items, including nested tuples and sets. For example `((((((((1,)*20,)*20,)*20,)*20,)*20,)*20,)*20,)*20` contains around `21**8` items.

The size of every tuple and set is limited to 256 (was 20).

What the limit looks reasonable to you?

Reply:
It's a matter of tradeoff between CPU (compute at runtime) vs disk space/memory.

If the optimized tuple is in a function which is almost never called, it would be waste to allocate 1000000 in memory, whereas it's never used.

IHMO the maximum size of "a constant" (a constant could be a tuple with nested tuples) should be 4 KB.

>>> sys.getsizeof((None,)*500)
4048

So I suggest to limit to 500 items, not 1000000.

Reply:
500 is too small. A tuple of 256 pairs contains 256*3 items total. I have decreased the limit to 1024.

Reply:
Why do you want to build a constant for a tuple of 768 items? What is the use case? Is it a performance issue to build it at runtime?

Reply:
Constant tuples or strings of size 256 are used in decoders. Constant dicts of tens elements are used in encoders. In sre_compile.py there is a constant expression `b'0' + b'1' * 255`. I think the size 256 is the reasonable limit of the size of constant collections. If the collection contains not primary types, but tuples, the total number of items is multiple times the size of the collection.

Using too small limit can cause a regression. Note that currently there is no limit. 1024 is much lesser than infinity.